### PR TITLE
add lisht kernel

### PR DESCRIPTION
**Pull request by WindQAQ**
- Address #437 - lisht kernel [LiSHT: Non-Parametric Linearly Scaled Hyperbolic Tangent Activation Function for Neural Networks](https://arxiv.org/abs/1901.05894)
- Remove the signature constraint for `tfa.activations.*`
- Have to find some ways to eliminated duplicated testing method

### CODE REVIEW

File: tensorflow_addons/custom_ops/activations/cc/kernels/lisht_op_gpu.cu.cc [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+using GPUDevice = Eigen::GpuDevice;
+
+#define DEFINE_GPU_KERNELS(T)                   \
+  template struct functor::Lisht<GPUDevice, T>; \
+  template struct functor::LishtGrad<GPUDevice, T>;
+
+TF_CALL_GPU_NUMBER_TYPES(DEFINE_GPU_KERNELS);
+
+}  // end namespace addons
+}  // namespace tensorflow
```
Comment (facaiy) [inline tensorflow_addons/custom_ops/activations/cc/kernels/lisht_op_gpu.cu.cc]:
```suggestion
}  // end namespace tensorflow
```

File: tensorflow_addons/custom_ops/activations/cc/kernels/lisht_op.h [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+template <typename Device, typename T>
+void LishtGradOp<Device, T>::OperateNoTemplate(OpKernelContext* context,
+                                               const Tensor& g, const Tensor& a,
+                                               Tensor* output) {
+  functor::LishtGrad<Device, T> functor;
+  functor(context->eigen_device<Device>(), g.flat<T>(), a.flat<T>(),
+          output->flat<T>());
+}
+
+}  // namespace addons
```
Comment (facaiy) [inline tensorflow_addons/custom_ops/activations/cc/kernels/lisht_op.h]:
```suggestion
}  // end namespace addons
```

ditto

File: tensorflow_addons/activations/lisht_test.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                                    ("float32", np.float32),
+                                    ("float64", np.float64))
+    def test_gradients(self, dtype):
+        x = tf.constant([-3.0, -2.0, -1.0, 0.0, 1.0, 2.0, 3.0], dtype=dtype)
+
+        with tf.GradientTape(persistent=True) as tape:
+            tape.watch(x)
+            y_ref = _ref_lisht(x)
+            y = lisht(x)
+        grad_ref = tape.gradient(y_ref, x)
```
[the remark this answers is not in the document]
Comment (WindQAQ) [reply tensorflow_addons/activations/lisht_test.py]: HI Facai, yep, it's duplicated. I wonder which one do you prefer? Comparing gradient with reference function or with the theoretical one.

File: tensorflow_addons/activations/lisht_test.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+
+@test_utils.run_all_in_graph_and_eager_modes
+class LishtTest(tf.test.TestCase, parameterized.TestCase):
+    @parameterized.named_parameters(("float16", np.float16),
+                                    ("float32", np.float32),
+                                    ("float64", np.float64))
+    def test_lisht(self, dtype):
+        x = (np.random.rand(2, 3, 4) * 5.0 - 2.5).astype(dtype)
+        self.assertAllCloseAccordingToType(lisht(x), _ref_lisht(x))
```
[the remark this answers is not in the document]
Comment (WindQAQ) [reply tensorflow_addons/activations/lisht_test.py]: Because lisht is symmetric along y-axis, I tend to test against [-2.0, -1.0, 0.0, 1.0, 2.0]. Do you think this is strong enough? Thanks :D.

File: tensorflow_addons/custom_ops/activations/cc/kernels/lisht_op.cc [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+#include "tensorflow_addons/custom_ops/activations/cc/kernels/lisht_op.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/register_types.h"
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+
+namespace tensorflow {
+namespace addons {
+
+using CPUDevice = Eigen::ThreadPoolDevice;
```
[the remark this answers is not in the document]
Comment (WindQAQ) [reply tensorflow_addons/custom_ops/activations/cc/kernels/lisht_op.cc]: Just out of curiosity. Is there any difference between these two? I do suppose our ops are compiled with c++11 standard.

Quote from [cppreference](https://en.cppreference.com/w/cpp/language/type_alias).
>> There is no difference between a type alias declaration and typedef declaration.

File: tensorflow_addons/activations/lisht_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (facaiy) [reply tensorflow_addons/activations/lisht_test.py]: The theoretical one looks better for me, what do you think, Tzu-Wei?

File: tensorflow_addons/activations/lisht_test.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (facaiy) [reply tensorflow_addons/activations/lisht_test.py]: Sounds good :-)

File: tensorflow_addons/custom_ops/activations/cc/kernels/lisht_op.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (facaiy) [reply tensorflow_addons/custom_ops/activations/cc/kernels/lisht_op.cc]: Good question. Their difference is: the alias declaration is compatible with templates, whereas the C style typedef is not. My first thought is for consistency with tf core.

File: tensorflow_addons/activations/lisht_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (WindQAQ) [reply tensorflow_addons/activations/lisht_test.py]: Theoretical +1

File: tensorflow_addons/custom_ops/activations/cc/kernels/lisht_op.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (WindQAQ) [reply tensorflow_addons/custom_ops/activations/cc/kernels/lisht_op.cc]: Seems that core TF mixes two of them :-(
https://github.com/tensorflow/tensorflow/search?p=2&q=using+CPUDevice&unscoped_q=using+CPUDevice
https://github.com/tensorflow/tensorflow/search?q=typedef+Eigen&unscoped_q=typedef+Eigen

File: tensorflow_addons/activations/README.md [hunk 6]
```diff
@@ -4,25 +4,26 @@
...
+| sparsemax | sparsemax  | https://arxiv.org/abs/1602.02068 |
+| tanhshrink| tanhshrink |                                  |
 
 
 ## Contribution Guidelines
 #### Standard API
 In order to conform with the current API standard, all activations
 must:
  * Be a `tf.function`.
- * Have the signature `fn(input, axis=-1, name=None)`.
```
[the remark this answers is not in the document]
Comment (WindQAQ) [reply tensorflow_addons/activations/README.md]: I think it would be better to remove it (at least the style is consistent in the same submodule). But not sure if other operations in `image` and `text` etc should keep it.

[the remark this answers is not in the document]
Comment (facaiy) [reply tensorflow_addons/activations/README.md]: I prefer to keep `name` in other modules :-)

File: tensorflow_addons/activations/BUILD [hunk 7]
```diff
@@ -57,6 +58,19 @@ py_test(
     ],
 )
 
+py_test(
+    name = "lisht_test",
+    size = "medium",
```
Comment (facaiy) [inline tensorflow_addons/activations/BUILD]: small?

File: tensorflow_addons/activations/lisht.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+def lisht(x):
+    """LiSHT: Non-Parameteric Linearly Scaled Hyperbolic Tangent Activation Function.
+
+    Computes linearly scaled hyperbolic tangent (LiSHT): `x * tanh(x)`
+
+    See [LiSHT: Non-Parameteric Linearly Scaled Hyperbolic Tangent Activation Function for Neural Networks](https://arxiv.org/abs/1901.05894).
+
+    Args:
+        x: A `Tensor`. Must be one of the following types:
+            `float16`, `float32`, `float64`.
```
Comment (fsx950223) [inline tensorflow_addons/activations/lisht.py]: Should support bfloat16 too

File: tensorflow_addons/custom_ops/activations/cc/ops/lisht_op.cc [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+#include "tensorflow/core/framework/op.h"
+#include "tensorflow/core/framework/shape_inference.h"
+
+namespace tensorflow {
+namespace addons {
+
+REGISTER_OP("Addons>Lisht")
+    .Input("features: T")
+    .Output("activations: T")
+    .Attr("T: {half, float, double}")
```
Comment (fsx950223) [inline tensorflow_addons/custom_ops/activations/cc/ops/lisht_op.cc]: Should support bfloat16 too

File: tensorflow_addons/custom_ops/activations/cc/ops/lisht_op.cc [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    .SetShapeFn(shape_inference::UnchangedShape);
+
+REGISTER_OP("Addons>LishtGrad")
+    .Input("gradients: T")
+    .Input("features: T")
+    .Output("backprops: T")
+    .Attr("T: {half, float, double}")
```
Comment (fsx950223) [inline tensorflow_addons/custom_ops/activations/cc/ops/lisht_op.cc]: ditto

File: tensorflow_addons/custom_ops/activations/BUILD [hunk 11]
```diff
@@ -49,6 +49,28 @@ cc_library(
...
+        "-x cuda",
+        "-nvcc_options=relaxed-constexpr",
+        "-nvcc_options=ftz=true",
+    ]),
+    deps = [
+        "@local_config_tf//:libtensorflow_framework",
+        "@local_config_tf//:tf_header_lib",
+    ] + if_cuda_is_configured([
+        "@local_config_cuda//cuda:cuda_libs",
+        "@local_config_cuda//cuda:cuda_headers",
```
Comment (fsx950223) [inline tensorflow_addons/custom_ops/activations/BUILD]: Maybe there should be a bazel function wrap the cuda config

File: tensorflow_addons/custom_ops/activations/cc/ops/lisht_op.cc (hunk 9 above)
Comment (WindQAQ) [reply tensorflow_addons/custom_ops/activations/cc/ops/lisht_op.cc]: Would be a good feature if we could support bfloat16. I will investigate it later and make all acitvations support bfloat16 in the following PRs. Thanks for the suggestions!

File: tensorflow_addons/activations/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (WindQAQ) [reply tensorflow_addons/activations/README.md]: Sure, I will create another PR to clean up activation functions (name arg, duplicated tests etc.)
